### PR TITLE
Extra optimizations

### PR DESCRIPTION
**Pull request by pvasilek**
Draft proposal for extra decompiler optimizations: universe case, dead IFs.

### CODE REVIEW

**Comment by pvasilek** [comment]
Replaces
	switch (...) {
		case [long.MinValue..long.MaxValue]: { CODE; }
	}
with
CODE;

Also, draft for
 if (...) nop; [else nop;] -> nop;
 if (...) { x; } x; -> x;
 if (...) { x; } else { x; } -> x;
But incomplete, need someones review and fix.

File: ICSharpCode.Decompiler/IL/ControlFlow/ControlFlowSimplification.cs [hunk 1]
```diff
@@ -101,6 +102,40 @@ bool IsSimple(ILInstruction inst)
...
+			foreach (var block in function.Descendants.OfType<Block>()) {
+				for (int i = block.Instructions.Count - 1; i >= 0; i--) {
+					if (block.Instructions[i] is IfInstruction ifInst) {
+						if (IsEmpty(ifInst.TrueInst) && IsEmpty(ifInst.FalseInst)) {
+							block.Instructions[i].ReplaceWith(new Nop().WithILRange(ifInst));
+						} else {
+							var t = ifInst.TrueInst;
+							var f = ifInst.FalseInst;
+							var n = i < block.Instructions.Count - 1 ? block.Instructions[i + 1] : null;
+							if (t.ToString().CompareTo(f.ToString()) == 0) {
```
Comment (dgrunwald) [inline ICSharpCode.Decompiler/IL/ControlFlow/ControlFlowSimplification.cs]: `ILInstruction.ToString()` is only meant for debug output and not for comparisons.

**Comment by dgrunwald** [comment]
The decompiler isn't supposed to optimize the code. If the compiler is inefficient and generates redundant conditions, we don't want to hide those from the user.

I would be OK with these changes behind the "Remove dead code" option, but not in the default mode.

**Comment by pvasilek** [comment]
I think you are right and this one may be rejected to become a part of de4dot. Plain decompiler is useless against highly obfuscated assemblies.

**Comment by dgrunwald** [comment]
To handle obfuscated code we'd first need to fix #901 by making our bytecode importer (`ILReader`) more like the logic in the .NET runtime.
Unreachable code needs to be discarded before we analyze the stack types; so `ControlFlowSimplification` is too late for that.

More advanced de-obfuscation, unpacking, etc. are a non-goal for ILSpy. There's already de4dot for that.

**Pull request closed without merging**